libraries/AP_Motors/AP_MotorsHeli_Quad.cpp

```diff
@@ -0,0 +1,309 @@
+    for (uint8_t i=0; i<AP_MOTORS_HELI_QUAD_NUM_MOTORS; i++) {
+        rc_enable_ch(AP_MOTORS_MOT_1+i);
+        _servo[i] = SRV_Channels::get_channel_for(SRV_Channel::Aux_servo_function_t(SRV_Channel::k_motor1+i), CH_1+i);
+        if (!_servo[i]) {
+            return false;
+        }
+    }
+    
+    // set rotor servo range
+    _rotor.init_servo();
+
+    _flags.initialised_ok = true;
+
+    return true;
+}
+
+// output_test - spin a motor at the pwm value specified
+//  motor_seq is the motor's sequence number from 1 to the number of motors on the frame
+//  pwm value is an actual pwm value that will be output, normally in the range of 1000 ~ 2000
+void AP_MotorsHeli_Quad::output_test(uint8_t motor_seq, int16_t pwm)
+{
+    // exit immediately if not armed
+    if (!armed()) {
+        return;
+    }
+
+    // output to motors and servos
+    switch (motor_seq) {
+    case 1 ... AP_MOTORS_HELI_QUAD_NUM_MOTORS:
+        rc_write(AP_MOTORS_MOT_1 + (motor_seq-1), pwm);
+        break;
+    case AP_MOTORS_HELI_QUAD_NUM_MOTORS+1:
+        // main rotor
+        rc_write(AP_MOTORS_HELI_QUAD_RSC, pwm);
+        break;
+    default:
+        // do nothing
+        break;
+    }
+}
+
+// set_desired_rotor_speed
+void AP_MotorsHeli_Quad::set_desired_rotor_speed(float desired_speed)
+{
+    _rotor.set_desired_speed(desired_speed);
+}
+
+// calculate_armed_scalars
+void AP_MotorsHeli_Quad::calculate_armed_scalars()
+{
+    _rotor.set_ramp_time(_rsc_ramp_time);
+    _rotor.set_runup_time(_rsc_runup_time);
+    _rotor.set_critical_speed(_rsc_critical/1000.0f);
+    _rotor.set_idle_output(_rsc_idle_output/1000.0f);
+    _rotor.set_power_output_range(_rsc_power_low/1000.0f, _rsc_power_high/1000.0f, _rsc_power_high/1000.0f, 0);
+}
+
+// calculate_scalars
+void AP_MotorsHeli_Quad::calculate_scalars()
+{
+    // range check collective min, max and mid
+    if( _collective_min >= _collective_max ) {
+        _collective_min = AP_MOTORS_HELI_COLLECTIVE_MIN;
+        _collective_max = AP_MOTORS_HELI_COLLECTIVE_MAX;
+    }
+
+    _collective_mid = constrain_int16(_collective_mid, _collective_min, _collective_max);
+
+    // calculate collective mid point as a number from 0 to 1000
+    _collective_mid_pct = ((float)(_collective_mid-_collective_min))/((float)(_collective_max-_collective_min));
+
+    // calculate factors based on swash type and servo position
+    calculate_roll_pitch_collective_factors();
+
+    // set mode of main rotor controller and trigger recalculation of scalars
+    _rotor.set_control_mode(static_cast<RotorControlMode>(_rsc_mode.get()));
+    calculate_armed_scalars();
+}
+
+// calculate_swash_factors - calculate factors based on swash type and servo position
+void AP_MotorsHeli_Quad::calculate_roll_pitch_collective_factors()
+{
+    // assume X quad layout, with motors at 45, 135, 225 and 315 degrees
+    // order FrontRight, RearLeft, FrontLeft, RearLeft
+    const float angles[AP_MOTORS_HELI_QUAD_NUM_MOTORS] = { 45, 225, 315, 135 };
```

Comment:
I think that the rotor position of this aircraft is a rectangle rather than a square.
The setting is square.
Is this value not so problematic about flight?

Reply:
that just changes the tuning gains a bit

Reply:
+    // order FrontRight, RearLeft, FrontLeft, RearLeft

the last RearLeft is RearRight?